File: benchmarks/mandelbrot_bench_broadcast.jl
# Mandelbrot escape-time benchmark (broadcast) — Julia / sjulia 共通ソース
#
# mandelbrot_bench_for.jl のスカラー for ループ版に対する broadcast 版。
# `xs' .+ im .* ys` で複素グリッドを作り、エスケープ関数を `.` broadcast で
# 全点に適用する — sjulia の broadcast + Complex サポートを示すベンチマーク。
# checksum(エスケープ回数の総和)は同一パラメータなら for 版と同じ定義。
#
# サイズ:
#   1700×1360, maxiter=500 — upstream julia で ~0.5 秒(M系 macOS)。
#     Julia 本家 / sjulia VM / Python 3.14(numpy) で比較可能。
#
# AoT (juliars): #8790 解消後にビルド可能。`--minimal-prelude` で
#   生成 Rust が rustc を通るようになったが、checksum が Julia 本家と
#   数カウントずれる（broadcast 実装の丸め経路の違い、Issue #9659 と同系）。
#   タイミング比較には影響しない。
# Python 版(numpy 使用): mandelbrot_bench_broadcast.py

function mandelbrot_escape(c::ComplexF64, maxiter::Int64)::Int64
    z = 0.0 + 0.0im
    for k in 1:maxiter
        if abs2(z) > 4.0
            return k - 1
        end
        z = z * z + c
    end
    return maxiter
end

function mandelbrot_grid(width::Int64, height::Int64, maxiter::Int64)
    xs = range(-2.0, 1.0; length=width)
    ys = range(1.2, -1.2; length=height)
    C = xs' .+ im .* ys
    counts = mandelbrot_escape.(C, maxiter)
    sum(counts)
end

function run_one(w::Int64, h::Int64, m::Int64)
    t0 = time_ns()
    r = mandelbrot_grid(w, h, m)
    t1 = time_ns()
    println(w, "x", h, " maxiter=", m, " total=", r, " t=", (t1 - t0) / 1.0e9)
end

mandelbrot_grid(50, 40, 50)   # warmup (JIT / 特殊化)
run_one(1700, 1360, 500)

File: benchmarks/mandelbrot_bench_broadcast.py
# Mandelbrot escape-time benchmark (broadcast, numpy) — Python mirror of
# mandelbrot_bench_broadcast.jl. Builds the complex grid by broadcasting
# (xs[None, :] + 1j * ys[:, None]) and iterates the escape map on whole
# arrays with an active mask — numpy's idiomatic vectorized form of the same
# algorithm (numpy use is explicitly allowed for this variant).
#
# The escape count matches the scalar definition: the number of z-updates
# performed before |z|^2 first exceeds 4, capped at maxiter.
#
# Run: uv run --python 3.14 --no-project --with numpy benchmarks/mandelbrot_bench_broadcast.py

import time

import numpy as np


def mandelbrot_grid(width: int, height: int, maxiter: int) -> int:
    # Match Julia broadcast grid exactly: xs' .+ im .* ys where
    # xs = range(-2.0, 1.0; length=width) and ys = range(1.2, -1.2; length=height).
    # np.linspace endpoints differ slightly, so use the same explicit formulas.
    xs = -2.0 + 3.0 * np.arange(width, dtype=np.float64) / (width - 1)
    ys = 1.2 - 2.4 * np.arange(height, dtype=np.float64) / (height - 1)
    C = xs[None, :] + 1j * ys[:, None]

    z = np.zeros_like(C)
    counts = np.zeros(C.shape, dtype=np.int64)
    active = np.ones(C.shape, dtype=bool)
    for _ in range(maxiter):
        # Check escape before updating, matching Julia semantics:
        # Julia returns k-1 where k is the first iteration with |z|^2 > 4.
        active &= (z.real * z.real + z.imag * z.imag) <= 4.0
        if not active.any():
            break
        z[active] = z[active] * z[active] + C[active]
        counts[active] += 1
    return int(counts.sum())


def run_one(w: int, h: int, m: int) -> None:
    t0 = time.perf_counter_ns()
    r = mandelbrot_grid(w, h, m)
    t1 = time.perf_counter_ns()
    print(f"{w}x{h} maxiter={m} total={r} t={(t1 - t0) / 1.0e9}")


mandelbrot_grid(50, 40, 50)  # warmup
run_one(1700, 1360, 500)
